sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GoneAndRetryWithRetryPolicy.java

```diff
@@ -15,190 +14,347 @@
-            return Mono.just(ShouldRetryResult.noRetry());
-        } else if (exception instanceof GoneException &&
-            !request.isReadOnly() &&
-            BridgeInternal.hasSendingRequestStarted((CosmosException)exception)) {
+
+        return this.retryWithRetryPolicy.shouldRetry(exception)
+                                        .flatMap((retryWithResult) -> {
+
+            if (retryWithResult.shouldRetry) {
+                return Mono.just(retryWithResult);
+            }
+
+            return this.goneRetryPolicy.shouldRetry(exception)
+                .flatMap((goneRetryResult) -> {
+                    if (!goneRetryResult.shouldRetry) {
+                        logger.debug("Operation will NOT be retried. Exception:",
+                            exception);
+                        this.end = Instant.now();
+                    }
+
+                    return Mono.just(goneRetryResult);
+                });
+        });
+    }
+
+    private Duration getElapsedTime() {
+        Instant endSnapshot = this.end != null ? this.end : Instant.now();
+
+        return Duration.between(this.start, endSnapshot);
+    }
+
+    static class GoneRetryPolicy extends RetryPolicyWithDiagnostics {
+        private final static int DEFAULT_WAIT_TIME_IN_SECONDS = 30;
+        private final static int MAXIMUM_BACKOFF_TIME_IN_SECONDS = 15;
+        private final static int INITIAL_BACKOFF_TIME = 1;
+        private final static int BACK_OFF_MULTIPLIER = 2;
+
+        private final RxDocumentServiceRequest request;
+        private volatile int attemptCount = 1;
+        private volatile int attemptCountInvalidPartition = 1;
+        private volatile int currentBackoffSeconds = GoneRetryPolicy.INITIAL_BACKOFF_TIME;
+        private final Supplier<Duration> getElapsedTimeSupplier;
+        private final int waitTimeInSeconds;
+        private final LastRetryWithExceptionProvider lastRetryWithExceptionProvider;
+        //TODO once this is moved to IRetryPolicy, remove from here.
+        public final static Quadruple<Boolean, Boolean, Duration, Integer> INITIAL_ARGUMENT_VALUE_POLICY_ARG = Quadruple.with(false, false,
+            Duration.ofSeconds(60), 0);
+
+        public GoneRetryPolicy(
+            RxDocumentServiceRequest request,
+            Integer waitTimeInSeconds,
+            Supplier<Duration> getElapsedTimeSupplier,
+            LastRetryWithExceptionProvider lastRetryWithExceptionProvider) {
+
+            checkNotNull(request, "request must not be null.");
+            this.request = request;
+            this.getElapsedTimeSupplier = getElapsedTimeSupplier;
+            this.waitTimeInSeconds = waitTimeInSeconds != null ? waitTimeInSeconds : DEFAULT_WAIT_TIME_IN_SECONDS;
+            this.lastRetryWithExceptionProvider = lastRetryWithExceptionProvider;
+        }
+
+        private boolean isRetryableException(Exception exception) {
+            if (exception instanceof GoneException ||
```

Comment:
GoneException can be retried with certain conditions, and with certain conditions it can't be retried.

I see that you coded those conditions outside and the caller of `isRetryableException(exception)` do the additional check.
so there for some GoneException `isRetryableException(GoneEx)` may return true but the code may decide to not retry. that may make the naming a bit counter-intuitive.

I wonder if we should negate the logic and call this `isNotRetryableException`. as for "not" case we can be certain. thougth?

not a big deal either way.

Reply:
Good point - changed in next iteration